src/wasm/wasm-ir-builder.cpp

```diff
@@ -0,0 +1,921 @@
+/*
+ * Copyright 2023 WebAssembly Community Group participants
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include <cassert>
+
+#include "ir/names.h"
+#include "wasm-ir-builder.h"
+
+using namespace std::string_literals;
+
+namespace wasm {
+
+namespace {
+
+Result<> validateTypeAnnotation(HeapType type, Expression* child) {
+  if (child->type == Type::unreachable) {
+    return Ok{};
+  }
+  if (!child->type.isRef() ||
+      !HeapType::isSubType(child->type.getHeapType(), type)) {
+    return Err{"invalid reference type on stack"};
+  }
+  return Ok{};
+}
```

Comment:
It seems a little odd to me to have this here, as it feels more of a concern for the parser. That is, non-parser users don't want this (except maybe as an assertion).

Reply:
The benefit of performing validation here is that in the future the wat and binary parsers (which could both use this utility in principle) will be able to share this validation logic rather than duplicating it. Validation is done only in the `makeXYZ` methods; `visit` should already assume everything is valid.

Reply:
FWIW I decided to delete most of the validation out of the `makeXYZ` methods as well. The only validation left is validation of type annotations that can only be done once we know a node's children.